### PR TITLE
fragment-mono: init at 1.21

### PR DESCRIPTION
**Pull request by noahgitsham**
## Description of changes

Adds a new package for the Fragment Mono font.

It's already available in the google-fonts package, but nice to have it separated aswell (as seen in the roboto packages).

(This is my first nixpkgs PR, please let me know if I've made any mistakes)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -643,6 +643,8 @@ with pkgs;
 
   firefly-desktop = callPackage ../applications/misc/firefly-desktop { };
 
+  fragment-mono = callPackage ../data/fonts/fragment-mono { };
+
```
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:647]:
```suggestion
```

File: pkgs/data/fonts/fragment-mono/default.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
+{ lib, stdenvNoCC, fetchzip }:
```
Comment (AndersonTorres) [inline pkgs/data/fonts/fragment-mono/default.nix:1]: Move the file to `pkgs/by-name/fr/fragment-mono/package.nix`.

File: pkgs/data/fonts/fragment-mono/default.nix [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  installPhase = ''
+    runHook preInstall
+
+    install -Dm644 fonts/ttf/*.ttf -t $out/share/fonts/truetype
+
+    runHook postInstall
+  '';
+
+  meta = {
+    homepage = "https://github.com/weiweihuanghuang/${pname}";
```
Comment (AndersonTorres) [inline pkgs/data/fonts/fragment-mono/default.nix:21]:
```suggestion
    homepage = "https://github.com/weiweihuanghuang/fragment-mono";
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by noahgitsham** [comment]
@AndersonTorres Ah apologies forgot to rtfm on pkgs specifically, let me fix all this.

**Review by AndersonTorres — APPROVED** [review]

File: pkgs/by-name/fr/fragment-mono/package.nix [hunk 4]
```diff
@@ -0,0 +1,35 @@
+{
+  lib,
+  stdenvNoCC,
+  fetchzip,
+}:
+
+stdenvNoCC.mkDerivation rec {
```
Comment (ryand56) [inline pkgs/by-name/fr/fragment-mono/package.nix:7]:
```suggestion
stdenvNoCC.mkDerivation (finalAttrs: {
```

File: pkgs/by-name/fr/fragment-mono/package.nix [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  pname = "fragment-mono";
+  version = "1.21";
+
+  src = fetchzip {
+    url = "https://github.com/weiweihuanghuang/${pname}/releases/download/${version}/${pname}-${version}.zip";
```
Comment (ryand56) [inline pkgs/by-name/fr/fragment-mono/package.nix:12]:
```suggestion
    url = "https://github.com/weiweihuanghuang/fragment-mono/releases/download/${finalAttrs.version}/fragment-mono-${version}.zip";
```
#277994

File: pkgs/by-name/fr/fragment-mono/package.nix [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    runHook preInstall
+
+    install -Dm644 fonts/ttf/*.ttf -t $out/share/fonts/truetype
+
+    runHook postInstall
+  '';
+
+  meta = {
+    homepage = "https://github.com/weiweihuanghuang/fragment-mono";
+    description = "Helvetica Monospace Coding Font";
```
Comment (ryand56) [inline pkgs/by-name/fr/fragment-mono/package.nix:26]:
```suggestion
    changelog = "https://github.com/weiweihuanghuang/fragment-mono/releases/tag/${finalAttrs.version}";
    description = "Helvetica Monospace Coding Font";
```

File: pkgs/by-name/fr/fragment-mono/package.nix [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    longDescription = ''
+      Fragment Mono is a monospaced coding version of Helvetica created
+      by modifying and extending Nimbus Sans by URW Design Studio.
+    '';
+    license = lib.licenses.ofl;
+    platforms = lib.platforms.all;
+    maintainers = [ lib.maintainers.noahgitsham ];
+  };
+}
```
Comment (ryand56) [inline pkgs/by-name/fr/fragment-mono/package.nix:35]:
```suggestion
})
```

**Comment by noahgitsham** [comment]
Is there some OfBorg command someone/me has to run to get it to run/skip those two Darwin tests or is there just a long unskippable queue? This package has no passthru.tests which is what it says it's waiting to run.

**Comment by ryand56** [comment]
> Is there some OfBorg command someone/me has to run to get it to run/skip those two Darwin tests or is there just a long unskippable queue? This package has no passthru.tests which is what it says it's waiting to run.

It's a long unskippable queue due to having less Darwin machines and the fact that Darwin is slow and not able to be emulated that well.

**Comment by noahgitsham** [comment]
> Squash the commits according to contributing guidelines.

What's wrong with the commits? Are the last two not covered by the `etc` in `(pkg-name): (from -> to | init at version | refactor | etc)`

**Comment by AndersonTorres** [comment]
A newly-added package does not need a commit for each fixup:

https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions

> If you have commits pkg-name: oh, forgot to insert whitespace: squash commits in this case. 
> Use git rebase -i. See [Squashing Commits](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_squashing) for additional information.

**Comment by noahgitsham** [comment]
> A newly-added package does not need a commit for each fixup:

I see, could be clearer to write that explicitly in CONTRIBUTING.md, as I assumed `oh, forgot to insert whitespace` only meant formatting changes.

**Comment by AndersonTorres** [comment]
> I see, could be clearer to write that explicitly in CONTRIBUTING.md, as I assumed `oh, forgot to insert whitespace` only meant formatting changes.

The idea behind this is more general, and a bit hard to express in words.
The idea is that small fixups - including but not limited to formatting - do not need an entry in the commit history.

Further, there is another unwritten rule: the first time a package is included in Nixpkgs, it should be included as if it was done perfectly. 
Therefore, even big fixups like "add changelog link" or "add Python as dependency (thanks EggSpam for noticing this!)" should not appear in the commit history of a new package.

(Obviously there are exceptions, such as including a package that requires a modification in another existing package - but this is another story...)

**Review by AndersonTorres — APPROVED** [review]

**Comment by noahgitsham** [comment]
Don't mean to sound impatient, there's no rush, but I'm curious as to why this hasn't been merged yet.

Does each PR need multiple manual approvals? Or can I run the merge bot command to do it now?

**Comment by AndersonTorres** [comment]
> Don't mean to sound impatient, there's no rush, but I'm curious as to why this hasn't been merged yet.

Because no human with the power to push the button pushed it yet.

> 
> Does each PR need multiple manual approvals? 

No. It just needs a human with the power to push the button to push it.

> Or can I run the merge bot command to do it now?

The merge bot will not push the button in this case. 
I will not elaborate, but the merge bot merges only PRs that came from the auto-updater bot, and only in specific circumstances.

------------------------

Summarizing, you need to find a human with the power to push the button.

There are some places you can find them:

- Discourse:
  - If no one has seen your PR, ask for a review here first: https://discourse.nixos.org/t/prs-ready-for-review/3032/
  - If your PR was reviewed, ask for merge here: https://discourse.nixos.org/t/prs-already-reviewed/2617
- Matrix: https://app.element.io/#/room/#review-requests:nixos.org
- Discourse (unofficial but a lot of people uses it): https://discord.gg/BqdJJFER4H
- Knock on their doors - I never tried this tbh since I am not in Europe.

**Comment by noahgitsham** [comment]
Great, thanks for all the help! This feels like a very welcoming community.

**Review by h7x4 — APPROVED** [review]

**Pull request merged**